internal/cmd/connector_test.go

```diff
@@ -0,0 +1,269 @@
+			TrustedCertificate: serverOpts.TLS.Certificate,
+		},
+		Name:         "testing",
+		CACert:       types.StringOrFile(readFile(t, "testdata/pki/connector.crt")),
+		CAKey:        types.StringOrFile(readFile(t, "testdata/pki/connector.key")),
+		EndpointAddr: types.HostPort{Host: "127.0.0.1", Port: 55555},
+		Kubernetes: connector.KubernetesOptions{
+			AuthToken: "auth-token",
+			Addr:      kubeSrv.URL,
+			CA:        types.StringOrFile(certs.PEMEncodeCertificate(kubeSrv.Certificate().Raw)),
+		},
+		Addr: connector.ListenerOptions{
+			HTTPS:   "127.0.0.1:0",
+			Metrics: "127.0.0.1:0",
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	t.Cleanup(cancel)
+
+	runAndWait(ctx, t, func(ctx context.Context) error {
+		return connector.Run(ctx, opts)
+	})
+
+	// check destination has been registered
+	var destination *models.Destination
+	poll.WaitOn(t, func(t poll.LogT) poll.Result {
+		destination, err = data.GetDestination(srv.DB(), data.ByName("testing"))
+		switch {
+		case errors.Is(err, internal.ErrNotFound):
+			return poll.Continue("destination not registered")
+		case err != nil:
+			return poll.Error(err)
+		}
+		return poll.Success()
+	})
+
+	// check the destination was updated
+	expected := &models.Destination{
+		Model: models.Model{
+			ID:        anyUID,
+			CreatedAt: time.Now(),
+			UpdatedAt: time.Now(),
+		},
+		OrganizationMember: models.OrganizationMember{OrganizationID: srv.DB().OrganizationID()},
+		Name:               "testing",
+		UniqueID:           "4ebfd7dabeec5b37eafd20e3775f70ab86c7422036367d77d9bebfa03864e08b",
+		ConnectionURL:      "127.0.0.1:55555",
+		ConnectionCA:       opts.CACert.String(),
+		LastSeenAt:         time.Now(),
+		Version:            "99.99.99999",
+		Resources:          models.CommaSeparatedStrings{"default", "ns1", "ns2"},
+		Roles:              models.CommaSeparatedStrings{"admin", "view", "edit", "custom", "logs"},
+	}
+	assert.DeepEqual(t, destination, expected, cmpDestinationModel)
+
+	poll.WaitOn(t, func(t poll.LogT) poll.Result {
+		fakeKube.writesLock.Lock()
+		defer fakeKube.writesLock.Unlock()
+		if len(fakeKube.writes) >= 3 {
+			return poll.Success()
+		}
+		return poll.Continue("request count %d waiting for 3", len(fakeKube.writes))
+	})
+
+	// check kube bindings were updated
+	expectedWrites := []*http.Request{
+		{
+			Method: "PUT",
+			URL:    &url.URL{Path: "/apis/rbac.authorization.k8s.io/v1/clusterrolebindings/infra:view"},
+		},
+		{
+			Method: "PUT",
+			URL:    &url.URL{Path: "/apis/rbac.authorization.k8s.io/v1/namespaces/ns1/rolebindings/infra:admin"},
+		},
+		{
+			Method: "PUT",
+			URL:    &url.URL{Path: "/apis/rbac.authorization.k8s.io/v1/namespaces/ns1/rolebindings/infra:logs"},
+		},
+	}
+	sort.Slice(fakeKube.writes, func(i, j int) bool {
+		return fakeKube.writes[i].URL.Path < fakeKube.writes[j].URL.Path
+	})
+	assert.DeepEqual(t, fakeKube.writes, expectedWrites, cmpHTTPRequestShallow)
+
+	// TODO: check proxy is listening
```

Comment:
Checking the proxy requires a way to get the random assigned port, which requires a larger change to the connector. I'm planning on making that change along with splitting up the goroutines. It should be similar to what I did for the server in #1430